runtime/vm/xz64/UpcallThunkGen.cpp

```diff
@@ -0,0 +1,931 @@
+	return instructionSize;
+}
+
+/*
+ * Generates an instruction to add a 32-bit integer constant to the 32-bit value in R1
+ * Format: "AFI R1,I2"
+ * R1 - sourceRegister
+ * I2 - integerConstant
+ */
+static I_32
+AFI(I_8 *instructionPtr, I_32 sourceRegister, I_32 immediateInteger)
+{
+	I_32 instructionSize = 6;
+	I_32 opcode = 0xC2090000;
+	*((I_32 *)instructionPtr) = (opcode | ((sourceRegister & 0x0000000F) << 20) | ((immediateInteger & 0xFFFF0000) >> 16));
+	instructionPtr[4] = (I_8)((immediateInteger & 0xFF00) >> 8);
+	instructionPtr[5] = (I_8)(immediateInteger & 0xFF);
+	return instructionSize;
+
+}
+
+/*
+ * Generates a sequence of instructions to copy data from a Java MemorySegment Object to a C struct.
+ * The struct is smaller than or equal to 4096 bytes.
+ */
+static I_32
+generateSmallStructCopyInstructions(I_8 *instructionPtr, I_32 structSize, I_32 offsetToParameterArea, I_32 sourceRegister, I_32 destinationRegister)
+{
+	I_32 totalInstructionSize = 0;
+
+	const I_32 numLargeCopies = structSize / MAX_MVC_COPY_LENGTH;
+	const I_32 residue = structSize % MAX_MVC_COPY_LENGTH;
+
+	// Generate upto 16 MVC instructions to copy up 4096 bytes. Each MVC instruction in this sequence will copy 256 bytes.
+	for(int i = 0; i < numLargeCopies; i++) {
+		// MVC D1(L,B1),D2(B2)
+		// The 'L' field is 0-based
+		totalInstructionSize += MVC(instructionPtr + totalInstructionSize, i * MAX_MVC_COPY_LENGTH, MAX_MVC_COPY_LENGTH - 1, destinationRegister, i * MAX_MVC_COPY_LENGTH, sourceRegister);
+	}
+	// Generate an instruction to copy the last set of bytes (less than 256).
+	if (residue != 0) {
+		// MVC D1(L,B1),D2(B2)
+		totalInstructionSize += MVC(instructionPtr + totalInstructionSize, numLargeCopies * MAX_MVC_COPY_LENGTH, residue - 1, destinationRegister, numLargeCopies * MAX_MVC_COPY_LENGTH, sourceRegister);
+	}
+	return totalInstructionSize;
+}
+
+/*
+ * Generates a sequence of instructions to copy data from a Java MemorySegment Object to a C struct.
+ * The struct is larger than 4096 bytes.
+ */
+static I_32
+generateLargeStructCopyInstructions(I_8 *instructionPtr, I_32 structSize, I_32 offsetToParameterArea, I_32 sourceRegister, I_32 destinationRegister)
+{
+	I_32 totalInstructionSize = 0;
+
+	const I_32 numFullSlots = structSize / MAX_MVC_COPY_LENGTH;
+	const I_32 residue = structSize % MAX_MVC_COPY_LENGTH;
+	const I_32 maxLargeCopies = MAX_MVC_DISPLACEMENT / MAX_MVC_COPY_LENGTH;
+
+	const I_32 loopResidue = numFullSlots % maxLargeCopies;
+
+	// loopInstructionsLength = 16x MVC(6) + 2x LAY (6) + 1x AFI (6)
+	const I_32 loopInstructionsLength = 114; // size in bytes
+	I_32 destinationDisplacementInt = -1;
+	I_32 sourceDisplacementInt = -1;
+	I_32 currMVCLength = -1;
+
+	// Load loop counter into R3
+	totalInstructionSize += IILF(instructionPtr + totalInstructionSize, R3, numFullSlots - loopResidue);
+
+	// Generate a loop that will copy 4096 bytes at a time (i.e. 16 MVC instructions). The loop concludes where there
```

Comment:
`The loop concludes where...` may be better worded as `The loop concludes when...`

Reply:
This should now be fixed in the final squashed commit.